internal/app/app.go

```diff
@@ -41,21 +45,35 @@ type App struct {
 	busObject *proxyManagerBus
 }
 
+type options struct {
+	authorizer authorizerer
+}
+type option func(*options)
+
+type authorizerer interface {
+	IsSenderAllowed(context.Context, string, dbus.Sender) error
+}
+
 // Apply is a function called via D-Bus to apply the system proxy settings.
-func (b *proxyManagerBus) Apply(http, https, ftp, socks, no, mode string) *dbus.Error {
+func (b *proxyManagerBus) Apply(sender dbus.Sender, http, https, ftp, socks, no, mode string) *dbus.Error {
 	// Methods calls spin up separate goroutines, so ensure we don't run them in parallel
 	b.mu.Lock()
 	defer b.mu.Unlock()
 
-	b.running = true
-
-	log.Infof("Called Apply(%s, %s, %s, %s, %s, %s)", http, https, ftp, socks, no, mode)
+	log.Debugf("Sender %s called Apply(%q, %q, %q, %q, %q, %q)", sender, http, https, ftp, socks, no, mode)
 
-	p, err := proxy.New(b.ctx, http, https, ftp, socks, no, mode)
 	// We need to cancel the context in a deferred function to get the final
 	// state of the error variable, and to let the main thread know it's safe to quit.
+	var err error
 	defer func() { b.running = false; b.cancel(err) }()
```

Comment:
Rereading this, we didn’t consider multiple simultaneous calls. We should exit when there is no more apply() calls in progress. Most of the time, there will be only one, but we have multiple clients at the same time, and should quit when all returned (so we probably need a refcount of active connection).

We should return to each client if they could apply or not the proxy setting independantly.

Reply:
Sure, do you mind creating a card so that we don’t forget?